src/sync/scheduler.ts

```diff
@@ -0,0 +1,144 @@
+const getStatusKey = (type: TaskType) => `${type}Status`;
+
+// These numbers were obtained experimentally by syncing a large customer with 60K quota from GitHub
+const RATE_LIMIT_QUOTA_PER_TASK_RESERVE = 500;
+const MAX_NUMBER_OF_SUBTASKS = 100;
+// Coefficient to determine pool size of the selection for the subtasks
+const SUBTASKS_POOL_COEF = 10;
+
+/**
+ *
+ * @param otherTasks - the array will be shuffled after the call, BEWARE!
+ */
+const calculateTasksUsingGitHubRateLimitQuota = async (subscription: Subscription, logger: Logger, mainTask: Task, otherTasks: Task[]) => {
+	try {
+		const metrics = {
+			trigger: "ratelimit_check_backfill"
+		};
+		const rateLimitResponse = await (await createInstallationClient(subscription.gitHubInstallationId, subscription.jiraHost, metrics, logger, subscription.gitHubAppId)).getRateLimit();
+		const rateLimitData = rateLimitResponse.data;
+
+		const availQuota = Math.min(rateLimitData.resources.core.remaining, rateLimitData.resources.graphql.remaining);
+		// We need to reserve some quota for the main task, too
+		const availQuotaForSubtasks = Math.max(0, availQuota - RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+		const allowedSubtasks = Math.floor(availQuotaForSubtasks / RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+
+		const nSubTasks = Math.min(allowedSubtasks, MAX_NUMBER_OF_SUBTASKS);
+
+		logger.info({ nSubTasks, rateLimitData }, `Using subtasks: ${nSubTasks}`);
+
+		if (nSubTasks === 0) {
+			return [mainTask];
+		}
+
+		return [mainTask, ...(
+			otherTasks.slice(0, nSubTasks * SUBTASKS_POOL_COEF).sort(() => Math.random() - 0.5).slice(0, nSubTasks)
+		)];
+	} catch (err) {
+		logger.warn({ err }, "Cannot determine rate limit, return only main task");
+		return [mainTask];
+	}
+};
+
+const mapSyncStateToTasks = (tasks: TaskType[], syncState: RepoSyncState): Task[] => {
+	const mappedTasks: Task[] = [];
+	tasks.forEach(
+		(taskType) => {
+			if (!syncState[getStatusKey(taskType)] || syncState[getStatusKey(taskType)] === "pending") {
+				mappedTasks.push({
+					task: taskType,
+					repositoryId: syncState.repoId,
+					repository: {
+						id: syncState.repoId,
+						name: syncState.repoName,
+						full_name: syncState.repoFullName,
+						owner: { login: syncState.repoOwner },
+						html_url: syncState.repoUrl,
+						updated_at: syncState.repoUpdatedAt?.toISOString()
+					},
+					cursor: syncState[getCursorKey(taskType)] || undefined
+				});
+			}
+		}
+	);
+	return mappedTasks;
+};
+
+/**
+ *
+ * @param subscription
+ * @param targetTasks - the first task is always same/deterministic! The tail is random from not complete ones tasks.
+ * 											The caller can call this function after some delay again (e.g. SQS exponential backoff) and
+ * 											it is guaranteed they will receive same task as the very first one (a.k.a. "main task")
+ * @param logger
+ */
+export const getNextTasks = async (subscription: Subscription, targetTasks: TaskType[], logger: Logger): Promise<Task[]> => {
```

Comment:
/nit
Instead of `Promise<Task[]>`, maybe following can make code more readable.
```javascript
Promise<{
  mainTask: Task,
  otherTasks: Task[]
}>
```